lib/yggdrasil/player.ex

```diff
@@ -12,9 +13,8 @@ defmodule Yggdrasil.Player do
     Supervisor.add_player user_id, [owner_pid, push_msg]
   end
 
-  def run_cmd(user_id, cmd = %Message{type: :command}) do
-    player_pid = Registry.get_player user_id
-    GenServer.cast(player_pid, {:run_cmd, cmd})
+  def run_cmd(user_id, cmd) do
+    Command.execute cmd, user_id
```

Comment:
I think what put me off was the player being in the move command, and well we don't have anything else so you put what we have and the more I think through it the protocol would probably be the best way to dispatch for now. So sorry for the initial reaction, that thing I can't put my finger on about what I don't like I think is the lack of a room so it looks odd right now haha.

What I'm hung up on at the moment is that way execute is defined to take a cmd and user_id -- would there ever need to be more? So what I have mocked up below introduces a context object, and the context is started in the player command, and passed into the command.execute function.

What came to mind is the way Plug works and how it's the backbone of phoenix. Plug is centered around a connection and you have a large number of functions that center around taking in a connection mutating and returning and it's very composable.

I wonder if we can't take a cue from this and do something similar which would then make this context object the backbone of a command in a game, then we can write smaller function that compose together.

```elixir
defmodule Yggdrasil.Command.MoveCommand do
  defstruct exit: nil
  import Game.context

  defimpl Yggdrasil.Command do
    alias Yggdrasil.Message

    def execute(context, %MoveCommand{exit: direction}) do
      context
      |> assigns(exit: direction)
      |> Room.exit 
      |> Room.enter
    end
  end
end

defmodule Room do
  import Game.Context

  # looks in assigns for room to enter, and enters room
  # could handle spawning it if it's not started.
  # the room would talk to the user.
  def enter(ctx = %Context{}) do
    room_to_enter = ctx.assigns.enter
    # ... 
    done(ctx)
  end

  # exits current room would send message "You head .."
  # operates on context and the current room knows it's exits perhaps
  # and injects the rooms id into the assigns
  def exit(ctx = %Context{}) do
    room = find_current_room(ctx.user_id)
    exit_user(ctx.user_id, room) # runs some game logic to talk to player
    next_room =  find_exit_id(room, ctx.assigns.exit)
    assigns(ctx, enter: next_room)
  end

  # define all the helper functions for a room like find_current_room etc..
end

# player module

def run_cmd(user_id, cmd) do
  Game.Context.new
  |> add_user(user_id)
  |> Command.execute(cmd)
end
```